src/launcher/java/org/truffleruby/launcher/Launcher.java

```diff
@@ -204,29 +204,42 @@ public static void processArguments(
             List<String> args,
             boolean parseHelpEtc,
             boolean unknownOptionFails,
-            boolean allowedJVMOptions) {
+            boolean isNative) {
 
         try {
             config.setOption(OptionsCatalog.EXECUTION_ACTION, ExecutionAction.UNSET);
 
-            new CommandLineParser(args, config, true, false, parseHelpEtc).processArguments();
+            final CommandLineParser argumentCommandLineParser = new CommandLineParser(args, config, true, false, parseHelpEtc);
+            argumentCommandLineParser.processArguments();
 
             if (unknownOptionFails && !config.getUnknownArguments().isEmpty()) {
                 throw new CommandLineException("unknown option " + config.getUnknownArguments().get(0));
             }
 
             if (config.getOption(OptionsCatalog.READ_RUBYOPT)) {
-                new CommandLineParser(getArgsFromEnvVariable("RUBYOPT"),
-                        config, false, true, true).processArguments();
-                new CommandLineParser(getArgsFromEnvVariable("TRUFFLERUBYOPT"),
-                        config, false, false, true).processArguments();
+                final List<String> rubyoptArgs = getArgsFromEnvVariable("RUBYOPT");
+                final List<String> trufflerubyoptArgs = getArgsFromEnvVariable("TRUFFLERUBYOPT");
+                new CommandLineParser(rubyoptArgs, config, false, true, true).processArguments();
+                new CommandLineParser(trufflerubyoptArgs, config, false, false, true).processArguments();
+
+                if (isNative) {
+                    // Append options from ENV variables to args after last interpreter option, which makes sure that
+                    // maybeExec processes--(native|jvm)* options and filtering them out for new process
+                    // if exec is being called.
```

Comment:
Why are they appended?
Is it expected they have precedence over `-J/--jvm/--native` options passed as direct arguments?
I would think the opposite behavior is expected: arguments have priority over env vars in case of conflicts.
In other words, process all arguments as if they were passed like `ruby $RUBYOPT $TRUFFLERUBYOPT CLI_ARGS`.

Reply:
I would have expected `ruby $RUBYOPT $TRUFFLERUBYOPT CLI_ARGS` as well but ruby does `ruby CLI_ARGS $RUBYOPT $TRUFFLERUBYOPT` so do we for both env-var-defined options.

Reply:
Locally, at least for `-W` I see CLI args have precedence over RUBYOPT in MRI:
```
$ RUBYOPT=-W0 ruby -W2 -e 'p $VERBOSE'
true
```
Do you mean we did that ordering before in TruffleRuby and maybe inherited from JRuby?

Note for `JAVA_OPTS`, CLI args override them in the Bash launcher.

Reply:
I was testing `env RUBYOPT='-r a' ruby -r b -e p:a` which fails on b not a. We had `ruby CLI_ARGS $RUBYOPT $TRUFFLERUBYOPT` before in OptionParser which i think is inherited from JRuby. I think this is best to investigate in another PR.

Reply:
Since the ordering was already like that, then indeed it's best to stay consistent in this PR.